system/uuid.lisp

```diff
@@ -0,0 +1,124 @@
+;;;; Copyright (c) 2018 Henry Harrington <henry.harrington@gmail.com>
+;;;; This code is licensed under the MIT license.
+;;;; Copyright (c) 2020 Philip Mueller (phil.mueller@fittestbits.com)
+;;;; This code is licensed under the MIT license.
+
+;;======================================================================
+;; UUID Layout and Byte Order (per RFC 4122)
+;;
+;;     Field                      Data Type         Octet #
+;; time-low               unsigned 32 bit integer     0-3
+;; time-mid               unsigned 16 bit integer     4-5
+;; time-hi-and-version    unsigned 16 bit interger    6-7
+;; clock-seq-hi-and-
+;;     reserved           unsigned  8 bit integer       8
+;; clock-seq-low          unsigned  8 bit integer       9
+;; node                   unsigned 48 bit integer    10-15
+;;
+;; All of the field are read/written as big endian
+;;
+;;======================================================================
+
+(in-package :mezzano.uuid)
+
+(defun generate-uuid (&optional (result-type 'string))
+  "Generate version 4 UUID from pseudo-random numbers. Formats supported ~
+   are: list, vector or string."
+  ;; Set bits 6 & 7 of clock-seq-hi-and-reserved to 0 and 1 respectively
+  ;; Set the four most significant bits (12 - 15) of
+  ;; time-hi-and-version field to #b0100
+  (case result-type
+    (list
+     (let ((result NIL))
+       (dotimes (i 16)
+         (push (case i
+                 (9 (logior #x40 (random 16)))
+                 (7 (logior (random 64) #x80))
+                 (t (random 256)))
+               result))
+       result))
```

Comment:
This is a bit confusing because the list is constructed backwards. (loop for i below 16 collect ...)